tortoise/verifying.go

```diff
@@ -32,23 +33,23 @@ func (v *verifying) resetWeights(voted types.LayerID) {
 	}
 }
 
-func (v *verifying) countBallot(logger log.Log, ballot *ballotInfo) {
+func (v *verifying) countBallot(logger *zap.Logger, ballot *ballotInfo) {
 	start := time.Now()
 
 	prev := v.layer(ballot.layer.Sub(1))
 	counted := !(ballot.conditions.badBeacon ||
 		prev.opinion != ballot.opinion() ||
 		prev.verifying.referenceHeight > ballot.reference.height)
-	logger.With().Debug("count ballot in verifying mode",
-		ballot.layer,
-		ballot.id,
-		log.ShortStringer("ballot opinion", ballot.opinion()),
-		log.ShortStringer("local opinion", prev.opinion),
-		log.Bool("bad beacon", ballot.conditions.badBeacon),
-		log.Stringer("weight", ballot.weight),
-		log.Uint64("reference height", prev.verifying.referenceHeight),
-		log.Uint64("ballot height", ballot.reference.height),
-		log.Bool("counted", counted),
+	logger.Debug("count ballot in verifying mode",
+		zap.Uint32("lid", ballot.layer.Uint32()),
+		zap.Stringer("ballot", ballot.id),
+		log.ZShortStringer("ballot opinion", ballot.opinion()),
```

Comment:
Do we need the indirection with `log.ZShortStringer`? This does the same without the need to allocate additional objects just to print a short string:

```suggestion
		zap.String("ballot opinion", ballot.opinion().ShortString()),
```

Reply:
ballot.opinion().ShortString() allocates the string. so it is completly the opposite of what is needed

Reply:
I don't think it makes any difference, please correct me if my Benchmark is wrong:

```golang
import (
	"testing"

	"go.uber.org/zap"
	"go.uber.org/zap/zaptest"
)

type shortString struct {
	val string
}

func (s *shortString) ShortString() string {
	return s.val
}

func Benchmark_ShortString(b *testing.B) {
	s := &shortString{val: "test"}

	log := zaptest.NewLogger(b)

	for i := 0; i < b.N; i++ {
		log.Info("test", zap.String("val", s.ShortString()))
	}
}

func Benchmark_ZShortString(b *testing.B) {
	s := &shortString{val: "test"}

	log := zaptest.NewLogger(b)

	for i := 0; i < b.N; i++ {
		log.Info("test", ZShortStringer("val", s))
	}
}
```

Results:

```bash
Benchmark_ShortString-6           467226              5961 ns/op            1446 B/op         15 allocs/op
--- BENCH: Benchmark_ShortString-6
    logger.go:130: 2023-06-15T09:00:07.697Z     INFO    test    {"val": "test"}
        ... [output truncated]
Benchmark_ZShortString-6          503100              8767 ns/op            1432 B/op         16 allocs/op
--- BENCH: Benchmark_ZShortString-6
    logger.go:130: 2023-06-15T09:00:10.511Z     INFO    test    {"val": "test"}
        ... [output truncated]
PASS
ok      github.com/spacemeshos/go-spacemesh/log 7.271s
```

Reply:
it would even indicate that the indirection causes an additional allocation and slowdown

Reply:
that variable is not a string, but bytes - to make a short string it needs to be encoded with hex. therefore your example is not useful

Reply:
@fasmat
```go
log.Info("test", zap.String("val", s.ShortString()))
```
will call ` s.ShortString()` eagerly even if INFO is disabled.